pyop2/fusion/extended.py

```diff
@@ -0,0 +1,706 @@
+        if self._initialized:
+            return
+        Kernel._globalcount += 1
+
+        # We need to distinguish between the kernel name and the function name(s).
+        # Since /fusion.Kernel/ are, in general, collections of functions, the same
+        # function (which is itself associated a Kernel) can appear in different
+        # /fusion.Kernel/ objects, but possibly under a different name (to avoid
+        # name clashes)
+        self._name = "_".join([k.name for k in kernels])
+        self._function_names = {self.cache_key: self._name}
+
+        self._cpp = any(k._cpp for k in kernels)
+        self._opts = dict(flatten([k._opts.items() for k in kernels]))
+        self._include_dirs = list(set(flatten([k._include_dirs for k in kernels])))
+        self._ldargs = list(set(flatten([k._ldargs for k in kernels])))
+        self._headers = list(set(flatten([k._headers for k in kernels])))
+        self._user_code = "\n".join(list(set([k._user_code for k in kernels])))
+        self._attached_info = {'fundecl': None, 'attached': False}
+
+        # What sort of Kernel do I have?
+        if fused_ast:
+            # A single AST (as a result of soft or hard fusion)
+            self._ast = fused_ast
+            self._code = self._ast_to_c(fused_ast)
+        else:
+            # Multiple functions (AST or strings, as a result of tiling)
+            self._ast = None
+            self._code = self._multiple_ast_to_c(kernels)
+        self._kernels = kernels
+
+        self._initialized = True
+
+    def __iter__(self):
+        for k in self._kernels:
+            yield k
+
+    def __str__(self):
+        return "OP2 FusionKernel: %s" % self._name
+
+
+# API for fused parallel loops
+
+class ParLoop(sequential.ParLoop):
+
+    """The root class of non-sequential parallel loops."""
+
+    pass
+
+
+class FusionParLoop(ParLoop):
+
+    def __init__(self, kernel, iterset, *args, **kwargs):
+        self._it_space = kwargs['it_space']
+        super(FusionParLoop, self).__init__(kernel, iterset, *args, **kwargs)
+
+    def _build_itspace(self, iterset):
+        """
+        Bypass the construction of a new iteration space.
+
+        This avoids type checking in base.ParLoop._build_itspace, which would
+        return an error when the fused loop accesses arguments that are not
+        accessed by the base loop.
+        """
+        return self._it_space
+
+
+# API for tiled parallel loops
+
+class TilingIterationSpace(base.IterationSpace):
+
+    """A simple bag of :class:`IterationSpace` objects for a sequence of tiled
+    parallel loops."""
+
+    def __init__(self, all_itspaces):
+        self._iterset = [i._iterset for i in all_itspaces]
+        self._extents = [i._extents for i in all_itspaces]
+        self._block_shape = [i._block_shape for i in all_itspaces]
+        assert all(all_itspaces[0].comm == i.comm for i in all_itspaces)
+        self.comm = all_itspaces[0].comm
```

Comment:
Again, no superclass constructor called.

It looks like you're fighting really hard against the pyop2 interface to shoehorn stuff in.  We should note this going forward given any potential redesigns.